src/Workspaces/Core/Portable/Workspace/Solution/ProjectOrDocumentId.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal readonly struct ProjectOrDocumentId : IEquatable<ProjectOrDocumentId>
+    {
+        private readonly object? _projectOrDocumentId;
+
+        public ProjectOrDocumentId(ProjectId projectId)
+        {
+            _projectOrDocumentId = projectId ?? throw new ArgumentNullException(nameof(projectId));
+        }
+
+        public ProjectOrDocumentId(DocumentId documentId)
+        {
+            _projectOrDocumentId = documentId ?? throw new ArgumentNullException(nameof(documentId));
+        }
+
+        public bool IsDefault => _projectOrDocumentId is null;
+
+        public ProjectId ProjectId => DocumentId?.ProjectId ?? _projectOrDocumentId as ProjectId ?? throw new InvalidOperationException();
+
+        public DocumentId? DocumentId => _projectOrDocumentId as DocumentId;
```

Comment:
Maybe check for default before returning as DocumentId?

Reply:
Would that change anything?

Reply:
That would make it consistent with `ProjectId` property, i.e. throw if default.

Reply:
This is different though. ProjectId should be always available, DocumentId is optional.

Reply:
Maybe this type is trying to do too much. Perhaps it should be a simple sum type and the logic to get ProjectId from DocumentId should be left to the callers.

Reply:
> Perhaps it should be a simple sum type and the logic to get ProjectId from DocumentId should be left to the callers.

This would defeat most of the advantages that led me to create this PR.

Reply:
Would it? I see only one place where this is used.